File: puppy-rootfs-template/usr/lib/puppy/share/src/online-check-tray.py
#!/usr/bin/python3
#monitor online internet connectivity

import requests
import gi
import subprocess
import time
import gc

gi.require_version('Gtk', '3.0')
gi.require_version('AppIndicator3', '0.1')

from threading import Thread
from gi.repository import Gtk, GLib
from gi.repository import AppIndicator3 as AppIndicator

class OnlineIndicator:

    def __init__(self):
		
        self.check_url="http://www.msftncsi.com/ncsi.txt"
						
        self.last_status=""
        self.gc_tick=0
        self.status_tick=0

        self.indicator = AppIndicator.Indicator.new(
            "weather-overcast-symbolic",
            "weather-overcast-symbolic",
            AppIndicator.IndicatorCategory.APPLICATION_STATUS
        )
        
        self.indicator.set_title("Online Status")
        self.indicator.set_status(AppIndicator.IndicatorStatus.ACTIVE)
        self.indicator.set_menu(self.build_menu())
		
        #periodically check UFW status
        self.update_status()
        self.status_check_thread = Thread(target=self.start_periodic_check)
        self.status_check_thread.daemon = True
        self.status_check_thread.start()

    def build_menu(self):
		
        menu = Gtk.Menu()

        #Add quit option
        quit_item = Gtk.MenuItem(label="Quit")
        quit_item.connect('activate', self.quit)
        menu.append(quit_item)

        menu.show_all()
        
        return menu

    def quit(self, source):
		
        Gtk.main_quit()

    def get_online_status(self):
		
      try:
        response = requests.get(self.check_url, timeout=5)        
        return "online"       
      except requests.RequestException:
        return "offline"

    def update_status(self):
		
        status = self.get_online_status()
        
        if self.last_status != status:
          
          self.last_status=status
          
          if status == "online":
            self.indicator.set_icon_full("weather-overcast-symbolic","Connnected to the internet")
          else:
            self.indicator.set_icon_full("weather-severe-alert-symbolic", "No internet connection")

    def start_periodic_check(self):

        while True:
			
            self.status_tick+=1
			
            if self.status_tick==3:
               GLib.idle_add(self.update_status)
               self.status_tick=0
               
            self.gc_tick+=1
            
            if self.gc_tick==30:
               collected=gc.collect()
               self.gc_tick=0				
            
            time.sleep(10)

if __name__ == "__main__":
	
    indicator = OnlineIndicator()
    Gtk.main()

File: puppy-rootfs-template/usr/lib/puppy/share/src/online-check-tray.c
// monitor online internet connectivity
// compile: gcc -o online-check-tray `pkg-config --cflags --libs gtk+-3.0 appindicator3-0.1 glib-2.0 libcurl` online-check-tray.c

#include <gtk/gtk.h>
#include <libappindicator/app-indicator.h>
#include <glib.h>
#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>

typedef struct {
    AppIndicator *indicator;
    char *last_status;
    int gc_tick;
    int status_tick;
    const char *check_url;
} OnlineIndicator;

void update_status(OnlineIndicator *online);
void *start_periodic_check(void *online_ptr);
char *get_online_status(OnlineIndicator *online);
void quit(GtkWidget *widget, gpointer data);
GtkWidget *build_menu(OnlineIndicator *online);

int main(int argc, char **argv) {
    gtk_init(&argc, &argv);

    OnlineIndicator *online = g_new0(OnlineIndicator, 1);
    online->check_url = "http://www.msftncsi.com/ncsi.txt";
    online->last_status = g_strdup("");
    online->gc_tick = 0;
    online->status_tick = 0;

    // Initialize AppIndicator
    online->indicator = app_indicator_new("weather-overcast-symbolic", "weather-overcast-symbolic", APP_INDICATOR_CATEGORY_APPLICATION_STATUS);
    app_indicator_set_title(online->indicator, "Online Status");
    app_indicator_set_status(online->indicator, APP_INDICATOR_STATUS_ACTIVE);
    app_indicator_set_menu(online->indicator, GTK_MENU(build_menu(online)));

    // Periodically check online status
    update_status(online);

    // Start periodic check in a separate thread
    pthread_t status_check_thread;
    pthread_create(&status_check_thread, NULL, start_periodic_check, online);
    pthread_detach(status_check_thread);

    gtk_main();
    return 0;
}

GtkWidget *build_menu(OnlineIndicator *online) {
    GtkWidget *menu = gtk_menu_new();

    // Add quit option
    GtkWidget *quit_item = gtk_menu_item_new_with_label("Quit");
    g_signal_connect(quit_item, "activate", G_CALLBACK(quit), NULL);
    gtk_menu_shell_append(GTK_MENU_SHELL(menu), quit_item);

    gtk_widget_show_all(menu);
    return menu;
}

void quit(GtkWidget *widget, gpointer data) {
    gtk_main_quit();
}

char *get_online_status(OnlineIndicator *online) {
    CURL *curl;
    CURLcode res;
    char *status = g_strdup("offline");

    curl = curl_easy_init();
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, online->check_url);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
        curl_easy_setopt(curl, CURLOPT_NOBODY, 1L); // We don't need body, just the status

        res = curl_easy_perform(curl);
        if (res == CURLE_OK) {
            status = g_strdup("online");
        }

        curl_easy_cleanup(curl);
    }
    return status;
}

void update_status(OnlineIndicator *online) {
    char *status = get_online_status(online);

    if (g_strcmp0(online->last_status, status) != 0) {
        g_free(online->last_status);
        online->last_status = g_strdup(status);

        if (g_strcmp0(status, "online") == 0) {
            app_indicator_set_icon_full(online->indicator, "weather-overcast-symbolic", "Connected to the internet");
        } else {
            app_indicator_set_icon_full(online->indicator, "weather-severe-alert-symbolic", "No internet connection");
        }
    }

    g_free(status);
}

void *start_periodic_check(void *online_ptr) {
    OnlineIndicator *online = (OnlineIndicator *)online_ptr;

    while (1) {
        online->status_tick += 1;
        if (online->status_tick == 3) {
            g_idle_add((GSourceFunc)update_status, online);
            online->status_tick = 0;
        }

        online->gc_tick += 1;
        if (online->gc_tick == 30) {
            //g_collect();
            online->gc_tick = 0;
        }

        sleep(10);
    }
}
